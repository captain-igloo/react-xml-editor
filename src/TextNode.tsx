import * as React from 'react';

import { Actions, BubbleOptions } from './types';

interface Props {
    actions: Actions;
    element: string;
    id: string[];
    text: string;
}

export default class TextNode extends React.Component<Props> {
    private ref: React.RefObject<HTMLDivElement | null>;

    public constructor(props: Props) {
        super(props);
        this.onClick = this.onClick.bind(this);
        this.ref = React.createRef();
    }

    public render(): React.ReactNode {
        const { text } = this.props;
        return (
            <div
                className="textnode focusable"
                onClick={this.onClick}
                ref={this.ref}
            >
                <span className="value">{ text }</span>
            </div>
        );
    }

    private onClick(event: React.MouseEvent): void {
        const { actions, element, id, text } = this.props;
        const bubbleOptions: Partial<BubbleOptions> = {
            id,
            element,
            show: true,
            value: text,
        };
        if (this.ref.current) {
            const rect = this.ref.current.getBoundingClientRect();
            bubbleOptions.left = rect.left;
            bubbleOptions.top = rect.top;
        }
        actions.showBubble(bubbleOptions);
        event.stopPropagation();
    }
}
