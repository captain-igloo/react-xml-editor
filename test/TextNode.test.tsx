import {
    describe,
    expect,
    jest,
    test,
} from '@jest/globals';
import { fireEvent, render } from '@testing-library/react';
import * as React from 'react';

import TextNode from '../src/TextNode';

describe('TextNode component', () => {
    test('text node renders properly', () => {
        const { container } = render(
            <TextNode
                actions={{} as any}
                element="element"
                id={['id']}
                text="text"
            />,
        );
        expect(container).toMatchSnapshot();
    });

    test('Click should show bubble', () => {
        const showBubble = jest.fn();
        const { getByText } = render(
            <TextNode
                actions={{ showBubble } as any}
                element="element"
                id={['id']}
                text="text"
            />,
        );
        fireEvent.click(getByText('text'));
        expect(showBubble).toHaveBeenCalledWith({
            element: 'element',
            id: ['id'],
            left: 0,
            show: true,
            top: 0,
            value: 'text',
        });
    });
});
