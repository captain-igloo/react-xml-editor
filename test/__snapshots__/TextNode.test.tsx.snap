// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TextNode component text node renders properly 1`] = `
<div>
  <div
    class="textnode focusable"
  >
    <span
      class="value"
    >
      text
    </span>
  </div>
</div>
`;
